File: tokyo-x/src/core/tool-schema.ts
export type RiskTier = 0 | 1 | 2 | 3;

export type ToolCategory =
  | "file"
  | "terminal"
  | "browser"
  | "screen"
  | "voice"
  | "memory"
  | "network"
  | "notify";

export type JsonType = "string" | "number" | "boolean" | "array" | "object";

export interface JsonSchemaField {
  type: JsonType;
  description?: string;
  enum?: (string | number)[];
  items?: JsonSchemaField;
  properties?: Record<string, JsonSchemaField>;
  required?: string[];
}

export interface JsonSchemaObject extends JsonSchemaField {
  type: "object";
  properties: Record<string, JsonSchemaField>;
  required?: string[];
}

export interface ToolDefinition {
  name: string;
  version: string;
  description: string;
  category: ToolCategory;
  riskTier: RiskTier;
  enabled: boolean;
  inputSchema: JsonSchemaObject;
  outputSchema?: JsonSchemaObject;
}

export interface ValidationResult {
  ok: boolean;
  errors: string[];
}

function validateField(field: JsonSchemaField, value: unknown, path: string, errors: string[]): void {
  if (field.enum && !(field.enum as unknown[]).includes(value)) {
    errors.push(`${path}: must be one of ${field.enum.join(", ")}`);
    return;
  }
  switch (field.type) {
    case "string":
      if (typeof value !== "string") errors.push(`${path}: expected string`);
      break;
    case "number":
      if (typeof value !== "number" || Number.isNaN(value)) errors.push(`${path}: expected number`);
      break;
    case "boolean":
      if (typeof value !== "boolean") errors.push(`${path}: expected boolean`);
      break;
    case "array":
      if (!Array.isArray(value)) {
        errors.push(`${path}: expected array`);
      } else if (field.items) {
        value.forEach((item, i) => validateField(field.items as JsonSchemaField, item, `${path}[${i}]`, errors));
      }
      break;
    case "object":
      if (typeof value !== "object" || value === null || Array.isArray(value)) {
        errors.push(`${path}: expected object`);
        break;
      }
      validateObject(field, value as Record<string, unknown>, path, errors);
      break;
  }
}

function validateObject(
  field: JsonSchemaField,
  value: Record<string, unknown>,
  path: string,
  errors: string[]
): void {
  for (const key of field.required ?? []) {
    const v = value[key];
    if (v === undefined || v === null) errors.push(`${path}.${key}: missing required property`);
  }
  for (const [key, child] of Object.entries(field.properties ?? {})) {
    if (value[key] !== undefined) validateField(child, value[key], path ? `${path}.${key}` : key, errors);
  }
}

export function validateAgainstSchema(schema: JsonSchemaObject, input: unknown): ValidationResult {
  if (typeof input !== "object" || input === null || Array.isArray(input)) {
    return { ok: false, errors: ["input: expected object"] };
  }
  const errors: string[] = [];
  validateObject(schema, input as Record<string, unknown>, "", errors);
  return { ok: errors.length === 0, errors };
}

export function toolPromptDescription(tool: ToolDefinition): string {
  const params = Object.entries(tool.inputSchema.properties)
    .map(([k, v]) => `${k}:${v.type}${tool.inputSchema.required?.includes(k) ? "" : "?"}`)
    .join(", ");
  return `- ${tool.name} [${tool.category}, tier ${tool.riskTier}] ${tool.description} Args: ${params || "none"}`;
}

const TOOL_NAME_RE = /^[a-z][a-z0-9_]*(\.[a-z][a-z0-9_]*)+$/;

export class ToolRegistry {
  private tools = new Map<string, ToolDefinition>();

  register(tool: ToolDefinition): this {
    if (!TOOL_NAME_RE.test(tool.name)) throw new Error(`invalid tool name: ${tool.name}`);
    if (!tool.inputSchema || tool.inputSchema.type !== "object") {
      throw new Error(`tool ${tool.name} needs an object inputSchema`);
    }
    if (this.tools.has(tool.name)) throw new Error(`tool already registered: ${tool.name}`);
    this.tools.set(tool.name, tool);
    return this;
  }

  get(name: string): ToolDefinition | undefined {
    return this.tools.get(name);
  }

  has(name: string): boolean {
    return this.tools.has(name);
  }

  list(): ToolDefinition[] {
    return [...this.tools.values()].sort((a, b) => a.name.localeCompare(b.name));
  }

  listByCategory(category: ToolCategory): ToolDefinition[] {
    return this.list().filter((t) => t.category === category);
  }

  validateInput(name: string, input: unknown): ValidationResult {
    const tool = this.tools.get(name);
    if (!tool) return { ok: false, errors: [`unknown tool: ${name}`] };
    if (!tool.enabled) return { ok: false, errors: [`tool disabled: ${name}`] };
    return validateAgainstSchema(tool.inputSchema, input);
  }

  describeForLLM(): string {
    return this.list().map(toolPromptDescription).join("\n");
  }
}

File: tokyox/core/tool_schema.py
from __future__ import annotations
import json
import re
from typing import Any

from .types import ToolDefinition, JsonSchemaField, ToolCategory, RiskTier, ValidationResult


TOOL_NAME_RE = re.compile(r"^[a-z][a-z0-9_]*(\.[a-z][a-z0-9_]*)+$")


def validate_field(field: JsonSchemaField, value: Any, path: str, errors: list[str]) -> None:
    if field.enum and value not in field.enum:
        errors.append(f"{path}: must be one of {field.enum}")
        return
    if field.type == "string":
        if not isinstance(value, str):
            errors.append(f"{path}: expected string")
    elif field.type == "number":
        if not isinstance(value, (int, float)) or isinstance(value, bool):
            errors.append(f"{path}: expected number")
    elif field.type == "boolean":
        if not isinstance(value, bool):
            errors.append(f"{path}: expected boolean")
    elif field.type == "array":
        if not isinstance(value, list):
            errors.append(f"{path}: expected array")
        elif field.items:
            for i, item in enumerate(value):
                validate_field(field.items, item, f"{path}[{i}]", errors)
    elif field.type == "object":
        if not isinstance(value, dict):
            errors.append(f"{path}: expected object")
        else:
            validate_object(field, value, path, errors)


def validate_object(field: JsonSchemaField, value: dict[str, Any], path: str, errors: list[str]) -> None:
    for key in field.required or []:
        if key not in value:
            errors.append(f"{path}.{key}: missing required property")
    for key, child in (field.properties or {}).items():
        if key in value and value[key] is not None:
            validate_field(child, value[key], f"{path}.{key}" if path else key, errors)


def validate_against_schema(schema: JsonSchemaField, input_data: Any) -> ValidationResult:
    if not isinstance(input_data, dict):
        return ValidationResult(ok=False, errors=["input: expected object"])
    errors: list[str] = []
    validate_object(schema, input_data, "", errors)
    return ValidationResult(ok=len(errors) == 0, errors=errors)


def tool_prompt_description(tool: ToolDefinition) -> str:
    if not tool.input_schema or not tool.input_schema.properties:
        params = "none"
    else:
        params = ", ".join(
            f"{k}:{v.type}{'' if tool.input_schema.required and k in tool.input_schema.required else '?'}"
            for k, v in tool.input_schema.properties.items()
        )
    return f"- {tool.name} [{tool.category.value}, tier {tool.risk_tier.value}] {tool.description}. Args: {params}"


class ToolRegistry:
    def __init__(self):
        self._tools: dict[str, ToolDefinition] = {}

    def register(self, tool: ToolDefinition) -> "ToolRegistry":
        if not TOOL_NAME_RE.match(tool.name):
            raise ValueError(f"invalid tool name: {tool.name}")
        if not tool.input_schema or tool.input_schema.type != "object":
            raise ValueError(f"tool {tool.name} needs an object input_schema")
        if tool.name in self._tools:
            raise ValueError(f"tool already registered: {tool.name}")
        self._tools[tool.name] = tool
        return self

    def get(self, name: str) -> ToolDefinition | None:
        return self._tools.get(name)

    def has(self, name: str) -> bool:
        return name in self._tools

    def list(self) -> list[ToolDefinition]:
        return sorted(self._tools.values(), key=lambda t: t.name)

    def list_by_category(self, category: ToolCategory) -> list[ToolDefinition]:
        return [t for t in self.list() if t.category == category]

    def validate_input(self, name: str, input_data: Any) -> ValidationResult:
        tool = self._tools.get(name)
        if not tool:
            return ValidationResult(ok=False, errors=[f"unknown tool: {name}"])
        if not tool.enabled:
            return ValidationResult(ok=False, errors=[f"tool disabled: {name}"])
        return validate_against_schema(tool.input_schema, input_data)

    def describe_for_llm(self) -> str:
        return "\n".join(tool_prompt_description(t) for t in self.list())